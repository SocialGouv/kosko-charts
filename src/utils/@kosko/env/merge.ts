//

/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-explicit-any */

// HACK(douglasduteil): wrap internal @kosko/env/dist/merge.js
// As the paths API is internal (not exported) in @kosko/env
// This is code will wrap and expose it for our needs.
// see https://github.com/tommy351/kosko/blob/%40kosko/env%401.0.1/packages/env/src/merge.ts#L12-L18
// since https://github.com/tommy351/kosko/commit/6353c8559994b826c8056ad03a35332e88b9134a

// https://stackoverflow.com/a/48769843
type UnionToIntersection<T> = (T extends any ? (k: T) => void : never) extends (
  k: infer I
) => void
  ? I
  : never;

export const merge: <T extends any[]>(
  ...data: T
) => UnionToIntersection<T[number]> =
  // eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires
  require("@kosko/env/dist/merge.js").merge;
