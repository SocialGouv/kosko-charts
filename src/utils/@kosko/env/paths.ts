//

import { join } from "path";

// HACK(douglasduteil): wrap internal @kosko/env/dist/paths.js
// As the paths API is internal (not exported) in @kosko/env
// This is code will wrap and expose it for our needs.
// see https://github.com/tommy351/kosko/blob/%40kosko/env%400.5.2/packages/env/src/paths.ts#L9-L13
// since https://github.com/tommy351/kosko/commit/6353c8559994b826c8056ad03a35332e88b9134a

export const formatPath: (
  path: string,
  data: Record<string, unknown>
) => string =
  // eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires
  require(join(require.resolve("@kosko/env"), "../paths.js")).formatPath;
