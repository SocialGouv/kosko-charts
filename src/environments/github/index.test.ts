import env from "./index";

test.each([
  ["because of missing variables", undefined],
  [
    "because of empty strings",
    {
      GITHUB_RUN_ID: " ",
      GITHUB_JOB: " ",
      GITHUB_REPOSITORY: " ",
      GITHUB_SHA: " ",
      GITHUB_REF: " ",
      SOCIALGOUV_KUBE_INGRESS_BASE_DOMAIN: " ",
    },
  ],
])("should throw %s", (_: string, testEnv?: NodeJS.ProcessEnv) => {
  expect(() => env(testEnv)).toThrowErrorMatchingSnapshot();
});

const validEnv = {
  GITHUB_RUN_ID: "1234",
  GITHUB_JOB: "5678",
  GITHUB_SHA: "abcdef",
  GITHUB_REPOSITORY: "socialgouv/sample-next-app",
  SOCIALGOUV_KUBE_INGRESS_BASE_DOMAIN: "dev2.fabrique.social.gouv.fr",
};
test.each([
  ["the github global env", { ...validEnv }],
  [
    "tagged github global env",
    {
      ...validEnv,
      GITHUB_REF: "some-branch",
    },
  ],
  [
    "preproproduction github global env",
    {
      ...validEnv,
      GITHUB_REF: "v4.12.2",
      PREPRODUCTION: "true",
    },
  ],
  [
    "production github global env",
    {
      ...validEnv,
      GITHUB_REF: "v4.12.2",
      PRODUCTION: "true",
    },
  ],
])("should return %s", (_: string, testEnv?: NodeJS.ProcessEnv) => {
  expect(env(testEnv)).toMatchSnapshot();
});
