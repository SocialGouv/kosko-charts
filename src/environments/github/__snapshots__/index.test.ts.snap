// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return preproduction tagged github global env 1`] = `
Object {
  "branch": "vX.Y.Z",
  "branchSlug": "vx-y-z-25jxw5",
  "environment": "vx-y-z-25jxw5",
  "isPreProduction": true,
  "isProduction": false,
  "metadata": Object {
    "annotations": Object {
      "app.github.com/job": "xxxxxxx-job",
      "app.github.com/ref": "refs/tags/vX.Y.Z",
      "app.github.com/repo": "socialgouv/sample-next-app",
      "app.github.com/run": "12345",
      "app.github.com/sha": "8843083",
    },
    "domain": "dev2.fabrique.social.gouv.fr",
    "git": Object {
      "branch": "refs/tags/vX.Y.Z",
      "remote": "socialgouv/sample-next-app",
    },
    "labels": Object {
      "application": "sample-next-app",
      "cert": "wildcard",
      "component": "sample-next-app",
      "owner": "sample-next-app",
      "team": "sample-next-app",
    },
    "namespace": Object {
      "name": "sample-next-app-preprod",
    },
    "rancherId": "c-bd7z2:p-7ms8p",
    "subdomain": "sample-next-app-preprod",
  },
  "productionNamespace": "sample-next-app",
  "projectName": "sample-next-app",
  "registry": "ghcr.io/socialgouv/sample-next-app",
  "sha": "8843083edb7f873cad1d1420731a60773594ffae",
  "shortSha": "8843083",
  "tag": "X.Y.Z",
}
`;

exports[`should return production github global env 1`] = `
Object {
  "branch": "vX.Y.Z",
  "branchSlug": "vx-y-z-25jxw5",
  "environment": "vx-y-z-25jxw5",
  "isPreProduction": false,
  "isProduction": true,
  "metadata": Object {
    "annotations": Object {
      "app.github.com/job": "xxxxxxx-job",
      "app.github.com/ref": "refs/tags/vX.Y.Z",
      "app.github.com/repo": "socialgouv/sample-next-app",
      "app.github.com/run": "12345",
      "app.github.com/sha": "8843083",
    },
    "domain": "dev2.fabrique.social.gouv.fr",
    "git": Object {
      "branch": "refs/tags/vX.Y.Z",
      "remote": "socialgouv/sample-next-app",
    },
    "labels": Object {
      "application": "sample-next-app",
      "component": "sample-next-app",
      "owner": "sample-next-app",
      "team": "sample-next-app",
    },
    "namespace": Object {
      "name": "sample-next-app",
    },
    "rancherId": "c-bd7z2:p-7ms8p",
    "subdomain": "sample-next-app",
  },
  "productionNamespace": "sample-next-app",
  "projectName": "sample-next-app",
  "registry": "ghcr.io/socialgouv/sample-next-app",
  "sha": "8843083edb7f873cad1d1420731a60773594ffae",
  "shortSha": "8843083",
  "tag": "X.Y.Z",
}
`;

exports[`should return the github global env 1`] = `
Object {
  "branch": "e2e-branch",
  "branchSlug": "e2e-branch",
  "environment": "e2e-branch",
  "isPreProduction": false,
  "isProduction": false,
  "metadata": Object {
    "annotations": Object {
      "app.github.com/job": "xxxxxxx-job",
      "app.github.com/ref": "refs/heads/e2e-branch",
      "app.github.com/repo": "socialgouv/sample-next-app",
      "app.github.com/run": "12345",
      "app.github.com/sha": "8843083",
    },
    "domain": "dev2.fabrique.social.gouv.fr",
    "git": Object {
      "branch": "refs/heads/e2e-branch",
      "remote": "socialgouv/sample-next-app",
    },
    "labels": Object {
      "application": "sample-next-app",
      "cert": "wildcard",
      "component": "sample-next-app",
      "owner": "sample-next-app",
      "team": "sample-next-app",
    },
    "namespace": Object {
      "name": "sample-next-app-e2e-branch",
    },
    "rancherId": "c-bd7z2:p-7ms8p",
    "subdomain": "sample-next-app-e2e-branch",
  },
  "productionNamespace": "sample-next-app",
  "projectName": "sample-next-app",
  "registry": "ghcr.io/socialgouv/sample-next-app",
  "sha": "8843083edb7f873cad1d1420731a60773594ffae",
  "shortSha": "8843083",
  "tag": "sha-8843083edb7f873cad1d1420731a60773594ffae",
}
`;

exports[`should return the github global env with renovate prefixed env name 1`] = `
Object {
  "branch": "renovate/major-@jest-monorepo",
  "branchSlug": "renovate-major-jest-monorepo-3qv5n7",
  "environment": "renovate-major-jest-monorepo-3qv5n7",
  "isPreProduction": false,
  "isProduction": false,
  "metadata": Object {
    "annotations": Object {
      "app.github.com/job": "xxxxxxx-job",
      "app.github.com/ref": "refs/heads/renovate/major-@jest-monorepo",
      "app.github.com/repo": "socialgouv/sample-next-app",
      "app.github.com/run": "12345",
      "app.github.com/sha": "8843083",
    },
    "domain": "dev2.fabrique.social.gouv.fr",
    "git": Object {
      "branch": "refs/heads/renovate/major-@jest-monorepo",
      "remote": "socialgouv/sample-next-app",
    },
    "labels": Object {
      "application": "sample-next-app",
      "cert": "wildcard",
      "component": "sample-next-app",
      "owner": "sample-next-app",
      "team": "sample-next-app",
    },
    "namespace": Object {
      "name": "sample-next-app-renovate-major-jest-monorepo-3qv5n7",
    },
    "rancherId": "c-bd7z2:p-7ms8p",
    "subdomain": "sample-next-app-renovate-major-jest-monorepo-3qv5n7",
  },
  "productionNamespace": "sample-next-app",
  "projectName": "sample-next-app",
  "registry": "ghcr.io/socialgouv/sample-next-app",
  "sha": "8843083edb7f873cad1d1420731a60773594ffae",
  "shortSha": "8843083",
  "tag": "sha-8843083edb7f873cad1d1420731a60773594ffae",
}
`;

exports[`should return the github global env with slugify 1`] = `
Object {
  "branch": "domifa-commander-8.x",
  "branchSlug": "domifa-commander-8-x-4k3z09",
  "environment": "domifa-commander-8-x-4k3z09",
  "isPreProduction": false,
  "isProduction": false,
  "metadata": Object {
    "annotations": Object {
      "app.github.com/job": "xxxxxxx-job",
      "app.github.com/ref": "refs/heads/domifa-commander-8.x",
      "app.github.com/repo": "socialgouv/sample-next-app",
      "app.github.com/run": "12345",
      "app.github.com/sha": "8843083",
    },
    "domain": "dev2.fabrique.social.gouv.fr",
    "git": Object {
      "branch": "refs/heads/domifa-commander-8.x",
      "remote": "socialgouv/sample-next-app",
    },
    "labels": Object {
      "application": "sample-next-app",
      "cert": "wildcard",
      "component": "sample-next-app",
      "owner": "sample-next-app",
      "team": "sample-next-app",
    },
    "namespace": Object {
      "name": "sample-next-app-domifa-commander-8-x-4k3z09",
    },
    "rancherId": "c-bd7z2:p-7ms8p",
    "subdomain": "sample-next-app-domifa-commander-8-x-4k3z09",
  },
  "productionNamespace": "sample-next-app",
  "projectName": "sample-next-app",
  "registry": "ghcr.io/socialgouv/sample-next-app",
  "sha": "8843083edb7f873cad1d1420731a60773594ffae",
  "shortSha": "8843083",
  "tag": "sha-8843083edb7f873cad1d1420731a60773594ffae",
}
`;

exports[`should throw because of empty strings 1`] = `
"Wrong environment variables
required \\"GITHUB_SHA\\": \\" \\" should not be an empty string
required \\"GITHUB_REF\\": \\" \\" should not be an empty string
required \\"GITHUB_JOB\\": \\" \\" should not be an empty string
required \\"GITHUB_RUN_ID\\": \\" \\" should not be an empty string
required \\"GITHUB_REPOSITORY\\": \\" \\" should not be an empty string
required \\"SOCIALGOUV_BASE_DOMAIN\\": \\" \\" should not be an empty string
"
`;
